File: Sources/AnsiSaver/ArtSource.swift
import Foundation
#if canImport(FoundationNetworking)
import FoundationNetworking
#endif


public protocol ArtSource {
    func loadArtPaths(completion: @escaping ([String]) -> Void)
}

public final class FolderSource: ArtSource {

    private let folderPath: String

    public init(folderPath: String) {
        self.folderPath = folderPath
    }

    public func loadArtPaths(completion: @escaping ([String]) -> Void) {
        let fm = FileManager.default
        guard let contents = try? fm.contentsOfDirectory(atPath: folderPath) else {
            completion([])
            return
        }

        let ansiExtensions: Set<String> = ["ans", "ansi", "asc", "diz", "ice", "bin", "xb", "pcb", "adf"]

        let paths = contents
            .filter { name in
                let ext = (name as NSString).pathExtension.lowercased()
                return ansiExtensions.contains(ext)
            }
            .map { (folderPath as NSString).appendingPathComponent($0) }

        completion(paths)
    }
}

public final class PackSource: ArtSource {

    private let packURL: String

    public init(packURL: String) {
        self.packURL = packURL
    }

    public func loadArtPaths(completion: @escaping ([String]) -> Void) {
        let packName = extractPackName(from: packURL)

        PackFetcher.fetchFileList(packURL: packURL) { filenames in
            guard !filenames.isEmpty else {
                completion([])
                return
            }

            let queue = DispatchQueue(label: "com.lardissone.AnsiSaver.packSource")
            var localPaths: [String] = []
            let group = DispatchGroup()

            for filename in filenames {
                let localPath = Cache.ansPath(forPack: packName, file: filename)

                if Cache.exists(localPath) {
                    queue.sync { localPaths.append(localPath) }
                    continue
                }

                group.enter()
                PackFetcher.downloadFile(packURL: self.packURL, filename: filename, to: localPath) { success in
                    if success {
                        queue.sync { localPaths.append(localPath) }
                    }
                    group.leave()
                }
            }

            group.notify(queue: .main) {
                completion(localPaths)
            }
        }
    }

    private func extractPackName(from url: String) -> String {
        let trimmed = url.hasSuffix("/") ? String(url.dropLast()) : url
        return (trimmed as NSString).lastPathComponent
    }
}

public final class URLSource: ArtSource {

    private let fileURLs: [String]

    public init(fileURLs: [String]) {
        self.fileURLs = fileURLs
    }

    public func loadArtPaths(completion: @escaping ([String]) -> Void) {
        let queue = DispatchQueue(label: "com.lardissone.AnsiSaver.urlSource")
        var localPaths: [String] = []
        let group = DispatchGroup()

        for urlString in fileURLs {
            let localPath = Cache.urlCachePath(for: urlString)

            if Cache.exists(localPath) {
                queue.sync { localPaths.append(localPath) }
                continue
            }

            guard let url = URL(string: urlString) else { continue }

            group.enter()
            let task = URLSession.shared.dataTask(with: url) { data, _, error in
                if let data = data, error == nil {
                    Cache.write(data, to: localPath)
                    queue.sync { localPaths.append(localPath) }
                }
                group.leave()
            }
            task.resume()
        }

        group.notify(queue: .main) {
            completion(localPaths)
        }
    }
}

File: ansi_saver/art_source.py
from __future__ import annotations

from pathlib import Path
from urllib.request import Request, urlopen

from .cache import Cache
from .pack_fetcher import PackFetcher


class FolderSource:
    _EXTENSIONS = {"ans", "ansi", "asc", "diz", "ice", "bin", "xb", "pcb", "adf"}

    def __init__(self, folder_path: str) -> None:
        self.folder_path = Path(folder_path)

    def load_art_paths(self) -> list[str]:
        if not self.folder_path.exists() or not self.folder_path.is_dir():
            return []
        return [
            str(path)
            for path in self.folder_path.iterdir()
            if path.is_file() and path.suffix.lower().lstrip(".") in self._EXTENSIONS
        ]


class PackSource:
    def __init__(self, pack_url: str) -> None:
        self.pack_url = pack_url

    def load_art_paths(self) -> list[str]:
        pack_name = self._extract_pack_name(self.pack_url)
        filenames = PackFetcher.fetch_file_list(self.pack_url)
        if not filenames:
            return []

        local_paths: list[str] = []
        for filename in filenames:
            local_path = Cache.ans_path(pack=pack_name, file=filename)
            if Cache.exists(local_path):
                local_paths.append(str(local_path))
                continue
            success = PackFetcher.download_file(self.pack_url, filename, str(local_path))
            if success:
                local_paths.append(str(local_path))
        return local_paths

    @staticmethod
    def _extract_pack_name(url: str) -> str:
        trimmed = url[:-1] if url.endswith("/") else url
        return trimmed.rsplit("/", maxsplit=1)[-1]


class URLSource:
    def __init__(self, file_urls: list[str]) -> None:
        self.file_urls = file_urls

    def load_art_paths(self) -> list[str]:
        local_paths: list[str] = []
        for url in self.file_urls:
            local_path = Cache.url_cache_path(url)
            if Cache.exists(local_path):
                local_paths.append(str(local_path))
                continue
            try:
                req = Request(url, headers={"User-Agent": "ansi-saver-python/1.0"})
                with urlopen(req, timeout=30) as response:
                    if response.status < 200 or response.status > 299:
                        continue
                    data = response.read()
            except Exception:
                continue

            Cache.write(data, local_path)
            local_paths.append(str(local_path))
        return local_paths
